### PR TITLE
Complete benchmarking pipeline - dataset creation, training, evaluation

### PR DESCRIPTION
**Pull request by dewball345**
Benchmarking includes:
* generating dataset - so far have generated cars196. Had a memory problem loading the entire dataset in ram, also had an issue putting 23 gb total in storage, so rewrote it using tfrecords. However, supports a sharded npz implementation(where each shard is sub-sharded) and the original full npz.
* training with dataset - have trained `CircleLoss`, `PNLoss`, `MultiSimilarityLoss`
* evaluation - get results from `model.calibrate` as well as `model.evaluate_retrieval`

Benchmarking metrics are in json form and can be generated yourself by following README. My test results are in the README as well. 

HOWEVER, seems to be an issue with the code. I'm guessing that it has to do with the model architectures and not the losses. Almost looks like model is reset after each epoch, so thinking that somewhere gradients aren't getting passed/updated.

I'm going to try making a dummy model from scratch(like in MNIST tutorial) and see if that works.

### CODE REVIEW

**Comment by dewball345** [comment]
I think there's something wrong with the tests - they're saying `no module named nmslib`?

**Comment by dewball345** [comment]
@owenvallis @ebursztein  I noticed that when the cars196 dataset is used in the hello world example(https://keras.io/examples/vision/metric_learning_tf_similarity/#visualization), the results are the same. The training loss doesn't decrease, and accuracy metrics are very poor. All I did was replace the cifar10 with cars196 and do the resizing preprocessing. 

Colab link: https://colab.research.google.com/drive/1SePk8mJBJFOBYxhfvjaJKaVyrUJGJz-4?usp=sharing

**Comment by dewball345** [comment]
Actually its probably my dataset. I saw this section
> Using a sampler is required to ensure that each batch contains at least N samples of each classes incuded in each batch. Otherwise contrastive loss does not work properly as it can't compute positive distances.

I'm going to try using this.

**Comment by dewball345** [comment]
Closing this for now - i'll probably make an updated one?
